Min API Validation fails for some generic types

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Describe the bug

When the input of an API is a generic type, such as `Dictionary<string,string>`, the build fails with errors in the generated code.

```text
/workspaces/dotnet10-issue-repros/validation-generic-types/obj/Debug/net10.0/Microsoft.AspNetCore.Http.ValidationsGenerator/Microsoft.AspNetCore.Http.ValidationsGenerator.ValidationsGenerator/ValidatableInfoResolver.g.cs(80,24): error CS0106: The modifier 'new' is not valid for this item
/workspaces/dotnet10-issue-repros/validation-generic-types/obj/Debug/net10.0/Microsoft.AspNetCore.Http.ValidationsGenerator/Microsoft.AspNetCore.Http.ValidationsGenerator.ValidationsGenerator/ValidatableInfoResolver.g.cs(80,24): error CS1520: Method must have a return type
/workspaces/dotnet10-issue-repros/validation-generic-types/obj/Debug/net10.0/Microsoft.AspNetCore.Http.ValidationsGenerator/Microsoft.AspNetCore.Http.ValidationsGenerator.ValidationsGenerator/ValidatableInfoResolver.g.cs(83,25): error CS0106: The modifier 'new' is not valid for this item
/workspaces/dotnet10-issue-repros/validation-generic-types/obj/Debug/net10.0/Microsoft.AspNetCore.Http.ValidationsGenerator/Microsoft.AspNetCore.Http.ValidationsGenerator.ValidationsGenerator/ValidatableInfoResolver.g.cs(83,25): error CS1520: Method must have a return type
/workspaces/dotnet10-issue-repros/validation-generic-types/obj/Debug/net10.0/Microsoft.AspNetCore.Http.ValidationsGenerator/Microsoft.AspNetCore.Http.ValidationsGenerator.ValidationsGenerator/ValidatableInfoResolver.g.cs(99,210): error CS0246: The type or namespace name 'ValidationOptions' could not be found (are you missing a using directive or an assembly reference?)
/workspaces/dotnet10-issue-repros/validation-generic-types/obj/Debug/net10.0/Microsoft.AspNetCore.Http.ValidationsGenerator/Microsoft.AspNetCore.Http.ValidationsGenerator.ValidationsGenerator/ValidatableInfoResolver.g.cs(98,10): error CS9206: An interceptor cannot be declared in the global namespace.
/workspaces/dotnet10-issue-repros/validation-generic-types/obj/Debug/net10.0/Microsoft.AspNetCore.Http.ValidationsGenerator/Microsoft.AspNetCore.Http.ValidationsGenerator.ValidationsGenerator/ValidatableInfoResolver.g.cs(81,17): error CS0246: The type or namespace name 'type' could not be found (are you missing a using directive or an assembly reference?)
/workspaces/dotnet10-issue-repros/validation-generic-types/obj/Debug/net10.0/Microsoft.AspNetCore.Http.ValidationsGenerator/Microsoft.AspNetCore.Http.ValidationsGenerator.ValidationsGenerator/ValidatableInfoResolver.g.cs(82,17): error CS0246: The type or namespace name 'members' could not be found (are you missing a using directive or an assembly reference?)
/workspaces/dotnet10-issue-repros/validation-generic-types/obj/Debug/net10.0/Microsoft.AspNetCore.Http.ValidationsGenerator/Microsoft.AspNetCore.Http.ValidationsGenerator.ValidationsGenerator/ValidatableInfoResolver.g.cs(84,25): error CS0246: The type or namespace name 'containingType' could not be found (are you missing a using directive or an assembly reference?)
/workspaces/dotnet10-issue-repros/validation-generic-types/obj/Debug/net10.0/Microsoft.AspNetCore.Http.ValidationsGenerator/Microsoft.AspNetCore.Http.ValidationsGenerator.ValidationsGenerator/ValidatableInfoResolver.g.cs(85,25): error CS0246: The type or namespace name 'propertyType' could not be found (are you missing a using directive or an assembly reference?)
/workspaces/dotnet10-issue-repros/validation-generic-types/obj/Debug/net10.0/Microsoft.AspNetCore.Http.ValidationsGenerator/Microsoft.AspNetCore.Http.ValidationsGenerator.ValidationsGenerator/ValidatableInfoResolver.g.cs(86,25): error CS0246: The type or namespace name 'name' could not be found (are you missing a using directive or an assembly reference?)
/workspaces/dotnet10-issue-repros/validation-generic-types/obj/Debug/net10.0/Microsoft.AspNetCore.Http.ValidationsGenerator/Microsoft.AspNetCore.Http.ValidationsGenerator.ValidationsGenerator/ValidatableInfoResolver.g.cs(87,25): error CS0246: The type or namespace name 'displayName' could not be found (are you missing a using directive or an assembly reference?)
```

The problem may come from the name mangling done for generic types, e.g. this line in the generated file `ValidatableInfoResolver.g.cs`:

```csharp
        private ValidatableTypeInfo CreateDictionary`2()
```

### Expected Behavior

Turning on validation should not cause build failures.

### Steps To Reproduce

The "validation-generic-types" project in this repo is a minimal repro of the problem:

https://github.com/mikekistler/dotnet10-issue-repros

The .http file contains a request that will trigger the error.

If you comment out the `AddValidations()` call Program.cs, the project builds and works correctly.

### Exceptions (if any)

_No response_

### .NET Version

10.0.100-preview.3.25201.16

### Anything else?

_No response_

--- a/src/Http/Http.Extensions/gen/Microsoft.AspNetCore.Http.ValidationsGenerator/Emitters/ValidationsGenerator.Emitter.cs
+++ b/src/Http/Http.Extensions/gen/Microsoft.AspNetCore.Http.ValidationsGenerator/Emitters/ValidationsGenerator.Emitter.cs
@@ -7,6 +7,7 @@ using Microsoft.CodeAnalysis.Text;
 using System.Text;
 using Microsoft.CodeAnalysis.CSharp;
 using System.IO;
+using System.Text.RegularExpressions;
 
 namespace Microsoft.AspNetCore.Http.ValidationsGenerator;
 
@@ -14,6 +15,7 @@ public sealed partial class ValidationsGenerator : IIncrementalGenerator
 {
     public static string GeneratedCodeConstructor => $@"global::System.CodeDom.Compiler.GeneratedCodeAttribute(""{typeof(ValidationsGenerator).Assembly.FullName}"", ""{typeof(ValidationsGenerator).Assembly.GetName().Version}"")";
     public static string GeneratedCodeAttribute => $"[{GeneratedCodeConstructor}]";
+    private static readonly Regex InvalidNameCharsRegex = new("[^0-9A-Za-z_]", RegexOptions.Compiled);
 
     internal static void Emit(SourceProductionContext context, (InterceptableLocation? AddValidation, ImmutableArray<ValidatableType> ValidatableTypes) emitInputs)
     {
@@ -238,12 +240,7 @@ namespace Microsoft.AspNetCore.Http.Validation.Generated
 
     private static string SanitizeTypeName(string typeName)
     {
-        // Replace invalid characters with underscores and remove generic notation
-        return typeName
-            .Replace(".", "_")
-            .Replace("<", "_")
-            .Replace(">", "_")
-            .Replace(",", "_")
-            .Replace(" ", "_");
+        // Replace invalid characters with underscores
+        return InvalidNameCharsRegex.Replace(typeName, "_");
     }
 }

--- a/src/Http/Http.Extensions/test/ValidationsGenerator/ValidationsGenerator.Parameters.cs
+++ b/src/Http/Http.Extensions/test/ValidationsGenerator/ValidationsGenerator.Parameters.cs
@@ -28,7 +28,7 @@ builder.Services.AddKeyedSingleton<TestService>("serviceKey");
 
 var app = builder.Build();
 
-app.MapGet("/params", (
+app.MapPost("/params", (
     // Skipped from validation because it is resolved as a service by IServiceProviderIsService
     TestService testService,
     // Skipped from validation because it is marked as a [FromKeyedService] parameter
@@ -39,7 +39,8 @@ app.MapGet("/params", (
     [CustomValidation(ErrorMessage = "Value must be an even number")] int value4 = 4,
     [CustomValidation, Range(10, 100)] int value5 = 10,
     // Skipped from validation because it is marked as a [FromService] parameter
-    [FromServices] [Range(10, 100)] int? value6 = 4) => "OK");
+    [FromServices] [Range(10, 100)] int? value6 = 4,
+    Dictionary<string, TestService>? testDict = null) => "OK");
 
 app.Run();
 

--- a/src/Http/Http.Extensions/test/ValidationsGenerator/snapshots/ValidationsGeneratorTests.CanValidateParameters#ValidatableInfoResolver.g.verified.cs
+++ b/src/Http/Http.Extensions/test/ValidationsGenerator/snapshots/ValidationsGeneratorTests.CanValidateParameters#ValidatableInfoResolver.g.verified.cs
@@ -67,6 +67,11 @@ namespace Microsoft.AspNetCore.Http.Validation.Generated
                 validatableInfo = CreateTestService();
                 return true;
             }
+            if (type == typeof(global::System.Collections.Generic.Dictionary<string, global::TestService>))
+            {
+                validatableInfo = CreateDictionary_2();
+                return true;
+            }
 
             return false;
         }
@@ -88,6 +93,38 @@ namespace Microsoft.AspNetCore.Http.Validation.Generated
                         propertyType: typeof(int),
                         name: "Value",
                         displayName: "Value"
+                    ),
+                ]
+            );
+        }
+        private ValidatableTypeInfo CreateDictionary_2()
+        {
+            return new GeneratedValidatableTypeInfo(
+                type: typeof(global::System.Collections.Generic.Dictionary<string, global::TestService>),
+                members: [
+                    new GeneratedValidatablePropertyInfo(
+                        containingType: typeof(global::System.Collections.Generic.Dictionary<string, global::TestService>),
+                        propertyType: typeof(global::System.Collections.Generic.ICollection<global::TestService>),
+                        name: "System.Collections.Generic.IDictionary<TKey,TValue>.Values",
+                        displayName: "System.Collections.Generic.IDictionary<TKey,TValue>.Values"
+                    ),
+                    new GeneratedValidatablePropertyInfo(
+                        containingType: typeof(global::System.Collections.Generic.Dictionary<string, global::TestService>),
+                        propertyType: typeof(global::System.Collections.Generic.IEnumerable<global::TestService>),
+                        name: "System.Collections.Generic.IReadOnlyDictionary<TKey,TValue>.Values",
+                        displayName: "System.Collections.Generic.IReadOnlyDictionary<TKey,TValue>.Values"
+                    ),
+                    new GeneratedValidatablePropertyInfo(
+                        containingType: typeof(global::System.Collections.Generic.Dictionary<string, global::TestService>),
+                        propertyType: typeof(global::TestService),
+                        name: "this[]",
+                        displayName: "this[]"
+                    ),
+                    new GeneratedValidatablePropertyInfo(
+                        containingType: typeof(global::System.Collections.Generic.Dictionary<string, global::TestService>),
+                        propertyType: typeof(global::System.Collections.ICollection),
+                        name: "System.Collections.IDictionary.Values",
+                        displayName: "System.Collections.IDictionary.Values"
                     ),
                 ]
             );

